syntax = "proto3";
option cc_enable_arenas = true;

package Ydb.S3Internal;
option java_package = "com.yandex.ydb.s3_internal";
option java_outer_classname = "S3InternalProtos";

import "ydb/public/api/protos/ydb_operation.proto";
import "ydb/public/api/protos/ydb_value.proto";

message S3ListingRequest {
    string table_name = 1;
    TypedValue key_prefix = 2; // A tuple representing all key columns that preceed path column
    string path_column_prefix = 3;
    string path_column_delimiter = 4;
    TypedValue start_after_key_suffix = 5; // A tuple representing key columns that succeed path column
    uint32 max_keys = 6;
    repeated string columns_to_return = 7;

    Ydb.Operations.OperationParams operation_params = 8;
}

message S3ListingResponse {
    Ydb.Operations.Operation operation = 1;
}

message S3ListingResult {
    Ydb.ResultSet common_prefixes = 1;    // Every Contents row starts with key suffix with KeySuffixSize columns
    Ydb.ResultSet contents = 2;           // Every Contents row starts with key suffix with KeySuffixSize columns
    uint32 key_suffix_size = 3;           // Number of key columns starting from path and up to the end
}
